Initial architecture doc

ostromart opened the pull request.

ayj commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,226 @@
+[schema](https://github.com/istio/operator/blob/master/pkg/apis/istio/v1alpha2/values_types.go).
+1. [Controller](K8s_controller) code. The code comprises the K8s listener, webhook and logic for reconciling the cluster
+to an IstioControlPlane CR.
+1. [Manifest creation](Manifest_creation) code. Manifest rendering is done by overlaying user settings on top of a
```
"Render Kubernetes manifests for the Istio Control Plane deployments from the user provided IstioControlPlane API specification. This involves applying the user's settings onto the selected base profile before Helm template rendering. Afterwards any optional user provided overlay patches are applied to the rendered manifests".

ostromart replied:
I reworked this a bit to be clearer and avoid passive voice. lmkwyt

ayj commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,226 @@
+1. [Manifest creation](Manifest_creation) code. Manifest rendering is done by overlaying user settings on top of a
+selected profile and passing the resulting values to a renderer in the Helm library to create manifests. Further
+overlays on the created manifests may be applied if they are defined in the user settings.
+1. [CLI](CLI) code. CLI code shares the IstioControlPlane API with the controller, but allows manifests to be generated
```
We also probably want an operational status/check area. This would be used by the operator for self-healing and CRD status. It could also be exposed through the CLI, e.g. `istioctl check`.

ayj commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,226 @@
+The operator code uses the new Helm charts in the [istio/installer](https://github.com/istio/installer) repo. It is not
+compatible with the older charts in [istio/istio](https://github.com/istio/istio/tree/master/install/kubernetes/helm).
+See the istio/installer repo for details about the new charts and why they were created. Briefly, the new charts
+are intended to support more realistic production deployments of Istio and hitless upgrades with canarying.
```
I'm not sure "realistic" adds much here. Maybe "... intended to support production ready Istio deployments following industry best practices, e.g. canary'd rollouts"

ostromart replied:
Done.

ayj commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,226 @@
+
+Throughout the document, the following terms are used:
+
+- New API: The API directly defined in the
```
"New API" won't age well :) Can we use `IstioControlPlane` instead? We can add a version qualifier if necessary to different the API maturity.

ostromart replied:
Good point, done. I've changed references to new and old APIs throughout the doc.

ayj commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,226 @@
+These rules are expressed in code in the
+[name](https://github.com/istio/operator/blob/905dd84e868a0b88c08d95b7ccf14d085d9a6f6b/pkg/name/name.go#L38) package.
+
+### K8s settings
```
AFAICT, all of these settings affect the deployment. Are there any use cases for exposing non-deployment related k8s settings here?

ostromart replied:
Yes, we have HPA settings.

sdake commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,226 @@
+
+## K8s controller
+
+WIP.
```
i'd recommend TODO(ostromart) as these re easier to grep in the code base.

ostromart replied:
Done.

sdake commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,226 @@
+
+## Manifest creation
+
+Manifest rendering is a multi-step process, shown in the figure below.
```
please use `SVG` for all images, not png.

ostromart replied:
Done.

sdake commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,226 @@
+is passed to the Helm rendering library and used to render the charts. The rendered manifests are passed to the next
+step.
+1. Overlays in the user CR are applied to the rendered manifests. No values are ever defined in configuration profile
+CRs at this layer, so no merge is performed in this step.
```
this whole section refers to new api and old api.  Instead of referring to them by age refer to them by some other form.  new and old are hard to parse (what functioanlity is in the old what is in the new).  This is a naming problem :)

ostromart replied:
Done.

rlenglet replied:
Please use "IstioControlPlaneSpec API" consistently instead of "new API". There are still a few occurrences of "new API" in the doc.

rlenglet replied:
Also, there are still occurrences of "old" in the doc. To me, this implies that this functionality will get deprecated at some point. Is that correct? What is the plan to replace Helm values.yaml in the operator? It seems like it's still an important part of the manifest generation. If this won't get deprecated, don't call it "old".

sdake commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,226 @@
+
+## Migration tools
+
+WIP.
```
TODO(ostromart) plese.

ostromart replied:
Done.

sdake commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,226 @@
+Policy | Policy
+Telemetry | Telemetry
+Security | Citadel
+| | Node agent
```
are these not implemented?  if so, ight add a third field which is there state (Not implemented, alpha, beta, stable).

ostromart replied:
They are, I think the table layout was just not clear. Updated.

sdake commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,226 @@
+
+the resulting namespaces will be:
+
+|Component|Namespace|
```
md requirees spaces on each side of block in a header please change to `| Component | Namesapce |`

ostromart replied:
Done.

sdake commented on ARCHITECTURE.md:
````diff
@@ -0,0 +1,226 @@
+      enabled: false
+```
+
+will enable all components of the Security feature except citadel.
````
lowercase "Security".  I am not sure if citadel should be capitalized - might ask @rcaballeromx or @geeknoid.

ostromart replied:
Done.

sdake commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,226 @@
+
+Rather than defining selective mappings from parameters to fields in K8s resources, the new API contains a consistent
+K8s block for each Istio component. The available K8s settings are defined in
+[KubernetesResourcesSpec](https://github.com/istio/operator/blob/905dd84e868a0b88c08d95b7ccf14d085d9a6f6b/pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto#L411):
```
is this line number persistent?

ostromart replied:
Yes, because it's at a sha.

geeknoid commented:
Sorry to keep bugging about this kind of stuff but...

Any reason why this is not a doc in our team drive? We don't keep design material in our source repos in general.

ostromart commented:
@geeknoid, the purpose of this text is to explain the code for developers wishing to contribute rather than being a full design doc (which we do have in docs). I've changed the title to "code overview" to reflect that and added a ptr to the full design doc.

rlenglet commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,231 @@
+The operator code uses the new Helm charts in the [istio/installer](https://github.com/istio/installer) repo. It is not
+compatible with the older charts in [istio/istio](https://github.com/istio/istio/tree/master/install/kubernetes/helm).
+See the istio/installer repo for details about the new charts and why they were created. Briefly, the new charts
+are intended to support production ready deployments of Istio that follow best practises like canarying for upgrade.
```
practises -> practices

ostromart replied:
Done.

rlenglet commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,231 @@
+
+- `IstioControlPlaneSpec`: The API directly defined in the
+[IstioControlPlaneSpec proto](https://github.com/istio/operator/blob/master/pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto),
+including feature and component groupings, namespaces and enablement, and per-component K8s settings. The term new API
```
new API -> "new API" or _new API_

ostromart replied:
I've removed all references to the old and new API.

rlenglet commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,231 @@
+Component settings are those that necessarily refer to a particular Deployment or Service. For example, the number
+of Pilot replicas is clearly a component setting, because it refers to a component which is a Deployment in the
+cluster. Most K8s platform settings are necessarily component settings.
+The available features and the components that comprise it are as follows:
```
comprise it -> comprise each feature

ostromart replied:
Done.

rlenglet commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,231 @@
+`IstioControlPlaneSpec` defines functional settings at the feature level. Functional settings are those that performs some
+function in the Istio control plane without necessarily being tied to any one component that runs in a Deployment.
+Component settings are those that necessarily refer to a particular Deployment or Service. For example, the number
+of Pilot replicas is clearly a component setting, because it refers to a component which is a Deployment in the
```
Remove "clearly".

ostromart replied:
Done.

rlenglet commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,231 @@
+
+Features and components can be individually or collectively enabled or disabled. If a feature is disabled, all of its
+components are disabled, regardless of their component level enablement. If a feature is enabled, all of its components
+are enabled, unless they individually disabled. For example:
```
unless they -> unless they are

ostromart replied:
Done.

rlenglet commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,231 @@
+
+Features and components are defined in the
+[name](https://github.com/istio/operator/blob/905dd84e868a0b88c08d95b7ccf14d085d9a6f6b/pkg/name/name.go#L38) package.
+
```
One question I have at this point is whether the operator manages dependencies between features and/or between components. E.g. does it disable Pilot automatically if CRDs is not enabled?
There is no mention of dependency management in the rest of the doc, so I assume the answer is "no."
I think it would be great to add a line here, or in the Enablement section to clarify that explicitly.

ostromart replied:
The answer is it only validates dependencies but does not automatically mutate configs. I've added some wording to that effect in the Validations section. 
For mutating the config, I think it's something we can consider for the future. I think for now it's probably clear enough that we don't do such things without making an explicit statement about it.

rlenglet commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,231 @@
+### Enablement
+
+Features and components can be individually or collectively enabled or disabled. If a feature is disabled, all of its
+components are disabled, regardless of their component level enablement. If a feature is enabled, all of its components
```
component level -> component-level

ostromart replied:
Done.

rlenglet commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,231 @@
+## Manifest creation
+
+Manifest rendering is a multi-step process, shown in the figure below. ![rendering
+process](images/operator_render_flow.svg) The example in the figure shows the rendering being triggered by a CLI `iop`
```
The diagram mentions "remote release URL," so I assume that the operator has the ability to download Helm charts? How does it work? Where are the charts stored? Are they periodically refreshed? How? Is the operator's filesystem read-only?

If you don't intend to document this feature, remove it from the diagram.

ostromart replied:
The feature is still not fully baked but will be there in the release. I've put a TODO here for now.

sdake commented:
@ostromart reference https://github.com/istio/operator/pull/90 for an example of what prow checks should look like.  The way to get the prow checks and circleci checks in sync with this PR is to force push after a master rebase e.g.:
```
git fetch upstream
git rebase upstream/master
git push -f origin/branchname
```

A pictorial representation of what the checks should look like:

part 1:
<img width="699" alt="Screen Shot 2019-07-16 at 5 16 55 AM" src="https://user-images.githubusercontent.com/755849/61293636-03587e00-a789-11e9-9196-02c48c7601d1.png">

part 2:
<img width="695" alt="Screen Shot 2019-07-16 at 5 17 08 AM" src="https://user-images.githubusercontent.com/755849/61293693-208d4c80-a789-11e9-96c5-f8de336c924d.png">

Cheers
-steve

sdake commented:
> Sorry to keep bugging about this kind of stuff but...
> 
> Any reason why this is not a doc in our team drive? We don't keep design material in our source repos in general.

@geeknoid 
What would make sense here is to have the `CONTRIBUTING.md` guide as normal, and then (at the choice of the repository maintainers) have a `CONTRIBUTING_DETAILS.md` file or simply `DETAILS.md` file.  This doc isn't quite that, but its pretty close and I believe is the intent of @ostromart.

Cheers
-steve

sdake commented:
**NB** I am going to keep this PR open to reduce the amount of churn I do while testing the CI jobs in prow.

1. Plase do not merge
2. please ignore this PR unless you want to follow along
